src/collapse/bpftrace.rs

```diff
@@ -0,0 +1,213 @@
+//! Module containting the **bpftrace** implementation of `Frontend`.
+
+use std::io;
+use std::iter::Peekable;
+use std::mem;
+
+use super::Frontend;
+
+/// The bpftrace implementation of `Frontend`.
+#[derive(Clone, Debug, Default, Eq, PartialEq, Hash)]
+pub struct Bpftrace {
+    state: State,
+}
+
+impl Frontend for Bpftrace {
+    fn collapse<R, W>(&mut self, reader: R, mut writer: W) -> io::Result<()>
+    where
+        R: io::BufRead,
+        W: io::Write,
+    {
+        self.state = State::NotInStack;
+
+        // Buffer for when we need to write the ascii representation of a number.
+        // 39 is the length, in bytes, it would take to write the largest 128 bit positive integer;
+        let mut number_buffer: [u8; 39] = unsafe { mem::uninitialized() };
+
+        // Iterate over each line
+        for line in reader.lines() {
+            let line = line?;
+
+            let mut is_beginning_of_line = true;
+
+            // Iterate over all characters in the line
+            let mut chars = line.chars().peekable();
```

Comment:
Sorry, I only got around to reading the actual implementation now!

I think I'd like to see this be less character-oriented. Specifically, I think we can:

 - Use `==` for `@[`
 - Use `.starts_with("]: ")` for `]: 123901238`, and then use `rsplit_whitespace().next()` to get the number

I also think you're missing a `.trim()` for when you're extracting the function name for each stack?

Reply:
The advantage of a character implementation is where sure to look at each character only once. Wouldn't deviating from that make it less efficient (although perhaps simpler)? Happy to do whatever you'd like, of course.

Reply:
While it's true that it's a little more efficient, `==` and `.starts_with` are also pretty efficient, since they'll only look at at most as many characters as are in the pattern. Thus we'll only look at a very small number of characters twice! And it makes the code _much_ easier to follow (I think). So yes, if you have the time to change it, I'd appreciate that!

Reply:
Begrudgingly 👍 :)